### ISSUE TITLE
Version 2.3.0 causing "Algolia::AlgoliaUnreachableHostError: Unreachable hosts"

### ISSUE THREAD

**Issue by paulspringett** [algolia/algoliasearch-client-ruby#484]
- Algolia Client Version: 2.3.0
- Language Version: Ruby 2.7.4

### Description

Upgrading to version 2.3.0 causes the following error when trying to connect to Algolia:

```
Algolia::AlgoliaUnreachableHostError: Unreachable hosts
```

These warnings have also started appearing in console output:

```
Top level ::CompositeIO is deprecated, require 'multipart/post' and use `Multipart::Post::CompositeReadIO` instead!
Top level ::Parts is deprecated, require 'multipart/post' and use `Multipart::Post::Parts` instead!

/Users/paul/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/faraday-net_http-2.1.0/lib/faraday/adapter/net_http.rb:36: warning: already initialized constant Faraday::Adapter::NetHttp::NET_HTTP_EXCEPTIONS
/Users/paul/.asdf/installs/ruby/2.7.4/lib/ruby/gems/2.7.0/gems/faraday-0.17.5/lib/faraday/adapter/net_http.rb:12: warning: previous definition of NET_HTTP_EXCEPTIONS was here
```

### Steps To Reproduce

1. Install Algolia 2.3.0
2. Configure an Algolia client with a valid app ID, admin API key and index name
3. Call `#set_settings`, for example

Issue does not occur on version 2.2.2

I suspect it was this PR that introduced the issue? https://github.com/algolia/algoliasearch-client-ruby/pull/471/files

**Comment by DevinCodes** [comment]
Hi @paulspringett !

Thank you for reporting this issue, and sorry to hear you're running into this. 

I tried to replicate the issue, but everything seems to work fine on my side. Would it be possible for you to give us a minimal repository in which there's a setup that replicates this issue, please? Without it, we won't be able to look into this behavior since I'm not able to reproduce this issue on my side with the steps you provided.

Thank you in advance! 😄

**Comment by paulspringett** [comment]
Hey @DevinCodes thanks so much for the quick response!

I think I've tracked it down to an older version of Faraday that we're using within our app:

```ruby
# algolia_faraday_test.rb
require "bundler/inline"

gemfile do
  source "https://rubygems.org"

  ruby "3.1.0"
  gem "faraday", "0.17.4"
  gem "algolia", "2.3.0"
end

search_config = Algolia::Search::Config.new(
  application_id: ENV.fetch("ALGOLIA_APPLICATION_ID"),
  api_key: ENV.fetch("ALGOLIA_ADMIN_API_KEY")
)

client = Algolia::Search::Client.create_with_config(search_config)
index = client.init_index(ENV.fetch("ALGOLIA_INDEX_NAME"))

puts index.get_settings
```

```sh
$ ALGOLIA_APPLICATION_ID=2GR****** ALGOLIA_ADMIN_API_KEY=292************* ALGOLIA_INDEX_NAME=example ruby ./algolia_faraday_test.rb
```

Looks like the version constraints in the gem still allow `faraday <= 0.15` so we were able to install `algolia 2.3.0`.

Maybe the min faraday constraint needs locking down to a newer version?

**Comment by DevinCodes** [comment]
You're right, it seems like it needs at least version 1 of faraday in order to work as expected, as it relies on including `faraday-net_http_persistent`, which doesn't work well with version `0.x` .  I tried to work around this in the `faraday-v0-debug` branch, and it seems to work for me locally. Would it be possible for you to (temporarily) change your project's Gemfile to use the following as the constraint for Algolia gem, please?

```ruby
gem 'algolia', git: 'git@github.com:algolia/algoliasearch-client-ruby.git', branch: 'faraday-v0-debug'
```

If it works, I'll draft a new release for this today, which should fix your issue in v2.3.1 then 😄

**Comment by paulspringett** [comment]
@DevinCodes Can confirm the `faraday-v0-debug` branch is working for us 👍

**Comment by DevinCodes** [comment]
@paulspringett this is available in v2.3.1 https://rubygems.org/gems/algolia/versions/2.3.1 !

**Comment by paulspringett** [comment]
@DevinCodes Brilliant, thank you!
